### PR TITLE
Prevent "invalid token" from being blocking

### PR DESCRIPTION
**Pull request by br41nslug**
## Scope

What's changed:

- The verifySessionToken function now throws a "InvalidCredentials" error to be consistent with the rest of the authentication middleware
- The session cookie will now be deleted if it's found to be invalid allowing a frontend to send you back to the login screen.

## Potential Risks / Drawbacks

- This solution feels like the simplest one, instead of throwing an error and leaving the end user in an error state requiring manual intervention to recover (clearing cookies)

## Review Notes / Questions

- I would like to lorem ipsum
- This PR complements #22433 or #22457, one of which will tackle the underlying issue
- As a follow up to the security fix https://github.com/directus/directus/pull/22353 (https://github.com/directus/directus/security/advisories/GHSA-g65h-35f3-x2w3)

### CODE REVIEW

**Review by licitdev — COMMENTED** [review]
Should we add a reason on the login page similar to `/admin/login?reason=SESSION_EXPIRED`?

image

**Review by hanneskuettner — COMMENTED** [review]
Should we add a unit test since we now expect the cookies to be cleared?

**Review by hanneskuettner — APPROVED** [review]
Nice tests! 

LGTM :guitar:

File: api/src/middleware/authenticate.test.ts [hunk 1]
```diff
@@ -266,3 +276,79 @@ test('Sets accountability to user information when static token is used', async
 	expect(req.accountability).toEqual(expectedAccountability);
 	expect(next).toHaveBeenCalledTimes(1);
 });
+
+test('Invalid session token clears the cookie', async () => {
```
Comment (paescuj) [inline api/src/middleware/authenticate.test.ts:280]: To make clear that we still expect the response to fail when clearing the cookie

```suggestion
test('Invalid session token responds with error and clears the cookie', async () => {
```

File: api/src/middleware/authenticate.test.ts [hunk 2]
```diff
@@ -266,3 +276,79 @@ test('Sets accountability to user information when static token is used', async
...
+		where: vi.fn().mockReturnThis(),
+		first: vi.fn().mockResolvedValue(null),
+	} as unknown as Knex);
+
+	await expect(handler(req, res, next)).rejects.toEqual(new InvalidCredentialsError());
+	expect(res.clearCookie).toHaveBeenCalledTimes(1);
+	expect(next).toHaveBeenCalledTimes(0);
+});
+
+test('Invalid query token does not clear the session cookie', async () => {
```
Comment (paescuj) [inline api/src/middleware/authenticate.test.ts:318]:
```suggestion
test('Invalid query token responds with error but does not clear the session cookie', async () => {
```

File: api/src/middleware/authenticate.ts [hunk 3]
```diff
@@ -45,7 +50,18 @@ export const handler = async (req: Request, _res: Response, next: NextFunction)
 		return next();
 	}
 
-	req.accountability = await getAccountabilityForToken(req.token, defaultAccountability);
+	try {
+		req.accountability = await getAccountabilityForToken(req.token, defaultAccountability);
+	} catch (err) {
+		if (isDirectusError(err, ErrorCode.InvalidCredentials)) {
```
Comment (paescuj) [inline api/src/middleware/authenticate.ts:56]: We should also clear the cookie on `InvalidToken` error

```suggestion
		if (isDirectusError(err, ErrorCode.InvalidCredentials) || isDirectusError(err, ErrorCode.InvalidToken)) {
```

Comment (hanneskuettner) [reply api/src/middleware/authenticate.ts:56]: Afaik the invalid token error can only be thrown if the user tampered with the session cookie, or in the insanely small time window between cookie expiration in the browser and session token expectation in the backend. And I think it is fair to have a failed state here.

Comment (paescuj) [reply api/src/middleware/authenticate.ts:56]: No, if the `SECRET` changed for some reason, the invalid token error is also thrown. This should not lock out users.

**Pull request merged**